### PR TITLE
Filters list of SPMs by user, initiator, and date 179378748

### PR DESCRIPTION
**Pull request by kwiliarty**

### CODE REVIEW

**Comment by kwiliarty** [comment]
@elliot I didn't add filters to other types of HUD reports. I wasn't sure if we wanted to include that right off. We should be able to by just rendering the filter in the relevant report templates. Let me know.

**Comment by kwiliarty** [comment]
Or we could add the filter into the report wrapper partial so that we have it only in one place. Just a thought.

**Comment by eanders** [comment]
Yep, let's expose this filter everywhere, and if you can get it in one partial, that'd be great.

**Comment by eanders** [comment]
Let's move the filter to the right across from the pagination
Screen Shot 2021-09-08 at 6 39 04 PM
Also, I'm going to merge in `release-5` which adds the 2021 versions of HUD reports and has some changes to this view.

**Comment by eanders** [comment]
Ok, so the merge was a bit tricky, so I fixed up the positioning and one issue with unfiltered results.  I think we're good.

File: app/controllers/hud_reports/base_controller.rb [hunk 1]
```diff
@@ -78,7 +79,24 @@ def set_reports
...
+        @reports.automated
+      else
+        @reports.manual
+      end
+
+      if @view_filter.present?
+        filter_range = @view_filter[:start].to_date..(@view_filter[:end].to_date + 1.days)
+        @reports = @reports.where(created_at: filter_range)
+      end
+      # TODO: add view filter @view_filter
```
Comment (eanders) [inline app/controllers/hud_reports/base_controller.rb]: What's the intent of this TODO?

File: app/views/hud_reports/_view_filter.haml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+          .o-filtration__option-row
+            .input-group.date{ data: { provide: :datepicker }}
+              %input.form-control{ name: :start, type: :text, value: @view_filter[:start] }
+              .input-group-append
+                %button.icon-calendar.btn.btn-secondary{ type: 'button' }
+            .o-filtration__separator to
+            .input-group.date{ data: {provide: :datepicker }}
+              %input.form-control{ name: :end, type: :text, value: @view_filter[:end] }
+              %span.input-group-append
+                %button.icon-calendar.btn.btn-secondary{ type: 'button' }
```
Comment (eanders) [inline app/views/hud_reports/_view_filter.haml:28]: Let's move date above the other options.

File: app/views/hud_reports/_view_filter.haml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    .o-filtration__header
+      %h3.client__filter-header Filter report list
+
+    = form_tag(nil, method: 'get') do
+
+      .o-filtration__option.row
+        .col-sm-12
+          .o-filtration__option-row
+            - type_options = options_for_select({'Normal (on-demand)' => :manual, 'Background (automated)' => :automated}, @view_filter[:initiator])
+            = render 'inputs/select', field_name: :initiator, label: 'Report Type', options: type_options
```
Comment (eanders) [inline app/views/hud_reports/_view_filter.haml]: Let's label this `Run type`.

File: app/views/hud_reports/_view_filter.haml [hunk 4]
```diff
@@ -0,0 +1,50 @@
+:ruby
+  if @report.present?
+    path_for_clear = path_for_report(@report)
+  else
```
Comment (eanders) [inline app/views/hud_reports/_view_filter.haml]: I think this half of the `if` can go away now.  The recent UI cleanup means show doesn't list other reports, only history and measures/questions.   So, actually we'll need to replace the top half of the if  in the other scenario.

**Review by eanders — CHANGES_REQUESTED** [review]
This looks really good.  
Todo:
- [x] Confirm it works for all HUD reports (APR, CAPER, CE APR, PATH, DQ, SPM)
- [x] Make it work on both `history` and `questions` / `measures` paths
- [x] Minor cleanup of the filter view as noted below.

File: app/controllers/hud_reports/base_controller.rb (hunk 1 above)
Comment (kwiliarty) [reply app/controllers/hud_reports/base_controller.rb]: Just cruft
✅

File: app/views/hud_reports/_view_filter.haml (hunk 2 above)
Comment (kwiliarty) [reply app/views/hud_reports/_view_filter.haml:28]: ✅

File: app/views/hud_reports/_view_filter.haml (hunk 3 above)
Comment (kwiliarty) [reply app/views/hud_reports/_view_filter.haml]: ✅

File: app/views/hud_reports/_view_filter.haml (hunk 4 above)
Comment (kwiliarty) [reply app/views/hud_reports/_view_filter.haml]: ✅

**Comment by kwiliarty** [comment]
@eanders Perhaps this is ready now?

File: app/views/hud_reports/_view_filter.haml [hunk 5]
```diff
@@ -0,0 +1,51 @@
+:ruby
+  filter_class = if @active_filter then 'filter-active' else 'filter' end
+  report_version = link_params.try(:[], :filter).try(:[], :report_version)
+  if report_version
+    path_for_clear_view_filter = "#{path_for_clear_view_filter}?filter[report_version]=#{report_version}"
```
Comment (eanders) [inline app/views/hud_reports/_view_filter.haml]: This is bound to end in disaster.  `path_for_clear_view_filter` should take an argument, then just pass in `{filter: {report_version: report_version}}` and have it pass that along to the method it calls.  Manually building the query string won't work if `path_for_clear_view_filter` ever returns a url with existing params.

**Review by eanders — CHANGES_REQUESTED** [review]
The path for clearing the view filter needs some help (as noted below).  Also, since you didn't change any database related items, `structure.sql` should not be included in this diff.  Same goes for `package-lock.json`.

**Comment by kwiliarty** [comment]
I've posted a new approach for getting the path to clear the filters. I've also reset `package-lock.json` and `db/warehouse/structure.sql`. I've had to do that repeatedly while working on this issue. Not sure why, but sorry it slipped through at the end.

**Comment by kwiliarty** [comment]
I've created `report_version` as a helper method in the HUD reports base controller.

**Review by eanders — APPROVED** [review]

**Pull request merged**